cc/position.cc

```diff
@@ -424,6 +429,264 @@ float Position::CalculateScore(float komi) {
+//      black-enclosed regions in R.
+//    - Remove from R all black-enclosed regions with a surrounding stone in a
+//      chain not in X.
+//
+// Unconditionally alive chains are also called pass-alive because they cannot
+// be captured by the opponent even if that player always passes on their turn.
+// More details:
+//   https://senseis.xmp.net/?BensonsDefinitionOfUnconditionalLife
+std::array<Color, kN * kN> Position::CalculatePassAliveRegions(
+    Color color) const {
+  struct BensonGroup {
+    Coord first_stone = Coord::kInvalid;
+    std::set<Coord> liberties;
+    int num_vital_regions = 0;
+    bool is_pass_alive = false;
+  };
+
+  struct BensonRegion {
+    Coord first_stone = Coord::kInvalid;
+    std::set<Coord> empty_points;
+    std::set<Coord> other_color_points;
+    std::set<BensonGroup*> vital_for;
+    std::set<BensonGroup*> enclosing_groups;
+    bool is_pass_alive = false;
+  };
+
+  std::vector<std::unique_ptr<BensonRegion>> region_pool;
+  std::vector<std::unique_ptr<BensonGroup>> group_pool;
+  std::array<BensonRegion*, kN * kN> regions;
+  std::array<BensonGroup*, kN * kN> groups;
+
+  // Build the set of all groups.
+  board_visitor_->Begin();
+  for (int row = 0; row < kN; ++row) {
+    for (int col = 0; col < kN; ++col) {
+      Coord c(row, col);
+      if (stones_[c].color() != color || !board_visitor_->Visit(c)) {
+        continue;
+      }
+
+      auto g = absl::make_unique<BensonGroup>();
+      g->first_stone = c;
+      while (!board_visitor_->Done()) {
+        c = board_visitor_->Next();
+        groups[c] = g.get();
+
+        for (auto nc : kNeighborCoords[c]) {
+          auto ns = stones_[nc];
+          if (ns.empty()) {
+            g->liberties.insert(nc);
+          } else if (ns.color() == color) {
+            board_visitor_->Visit(nc);
+          }
+        }
+      }
+      group_pool.push_back(std::move(g));
+    }
+  }
+
+  // Build the set of all regions.
+  board_visitor_->Begin();
+  for (int row = 0; row < kN; ++row) {
+    for (int col = 0; col < kN; ++col) {
+      Coord c(row, col);
+      if (stones_[c].color() == color || !board_visitor_->Visit(c)) {
+        continue;
+      }
+
+      auto r = absl::make_unique<BensonRegion>();
+      r->first_stone = c;
+      while (!board_visitor_->Done()) {
+        c = board_visitor_->Next();
+
+        regions[c] = r.get();
+        if (stones_[c].empty()) {
+          r->empty_points.insert(c);
+        } else {
+          r->other_color_points.insert(c);
+        }
+
+        for (auto nc : kNeighborCoords[c]) {
+          auto ns = stones_[nc];
+          if (ns.color() != color) {
+            board_visitor_->Visit(nc);
+          } else {
+            r->enclosing_groups.insert(groups[nc]);
+          }
+        }
+      }
+
+      // Find the vital groups for this region.
+      for (auto* g : r->enclosing_groups) {
+        bool is_vital =
+            std::includes(g->liberties.begin(), g->liberties.end(),
```

Comment:
what a clumsy api for subset calculation... why doesn't this  just take g->liberties and r->empty_points as args? does it assume that the set members are sorted for O(N+M) runtime?

Reply:
All the C++ library algorithms APIs are like this, I have no idea why they never introduced some kind of `range` abstraction :(

`std::includes` does assume both ranges are sorted, which `std::set` guarantees (it's often implemented as a red-black tree).